onnxruntime/core/session/compile_api.cc

```diff
@@ -0,0 +1,242 @@
+#include "core/session/abi_session_options_impl.h"
+#include "core/session/inference_session.h"
+#include "core/session/model_compilation_options.h"
+#include "core/session/onnxruntime_session_options_config_keys.h"
+#include "core/session/ort_apis.h"
+#include "core/session/ort_env.h"
+#include "core/session/utils.h"
+
+#define ORT_C_API_RETURN_IF_ERROR(expr)                 \
+  do {                                                  \
+    auto _status = (expr);                              \
+    if ((!_status.IsOK())) return ToOrtStatus(_status); \
+  } while (0)
+
+using namespace onnxruntime;
+
+ORT_API(void, OrtCompileAPI::ReleaseModelCompilationOptions,
+        _Frees_ptr_opt_ OrtModelCompilationOptions* ort_model_compile_options) {
+  delete reinterpret_cast<onnxruntime::ModelCompilationOptions*>(ort_model_compile_options);
+}
+
+ORT_API_STATUS_IMPL(OrtCompileAPI::CreateModelCompilationOptionsFromSessionOptions, _In_ const OrtEnv* env,
+                    _In_ const OrtSessionOptions* session_options, _Outptr_ OrtModelCompilationOptions** out) {
+  API_IMPL_BEGIN
+  if (env == nullptr) {
+    return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "The env argument must be a non-null pointer");
+  }
+
+  if (session_options == nullptr) {
+    return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "The session_options argument must be a non-null pointer");
+  }
+
+  auto model_compile_options = std::make_unique<onnxruntime::ModelCompilationOptions>();
+  model_compile_options->session_options = std::make_unique<OrtSessionOptions>(*session_options);
+
+  model_compile_options->session_options->value.has_explicit_ep_context_gen_options = true;
+  model_compile_options->session_options->value.ep_context_gen_options = session_options->value.GetEpContextGenerationOptions();
+  model_compile_options->session_options->value.ep_context_gen_options.enable = true;
+  model_compile_options->session_options->value.ep_context_gen_options.overwrite_existing_output_file = true;
+  model_compile_options->session_options->value.ep_context_gen_options.error_if_no_compiled_nodes = true;
+  ORT_C_API_RETURN_IF_ERROR(model_compile_options->session_options->value.config_options.AddConfigEntry(kOrtSessionOptionEpContextEnable, "1"));
+  *out = reinterpret_cast<OrtModelCompilationOptions*>(model_compile_options.release());
+  return nullptr;
+  API_IMPL_END
+}
+
+ORT_API_STATUS_IMPL(OrtCompileAPI::ModelCompilationOptions_SetInputModelPath,
+                    _In_ OrtModelCompilationOptions* ort_model_compile_options,
+                    const ORTCHAR_T* input_model_path) {
+  API_IMPL_BEGIN
+  auto model_compile_options = reinterpret_cast<onnxruntime::ModelCompilationOptions*>(ort_model_compile_options);
+
+  // Clear settings related to the input model. We want to allow the user to, for example, say the input model
+  // comes from a buffer and then change their mind and say it comes from file.
+  model_compile_options->ResetInputModelSettings();
```

Comment:
in general should we do validation of the args for the function before we reset anything? i.e. current values are preserved until we know we have new valid ones to use.

Reply:
Good call.

Reply:
Moved these Reset calls to appear after validation.